### PR TITLE
Add a `try_clone()` function to `OwnedFd`.

### PR DESCRIPTION
**Pull request by sunfishcode**
As suggested in #88564. This adds a `try_clone()` to `OwnedFd` by
refactoring the code out of the existing `File`/`Socket` code.

r? @joshtriplett

### CODE REVIEW

File: library/std/src/sys/unix/fd.rs [hunk 1]
```diff
@@ -266,22 +266,9 @@ impl FileDesc {
         }
     }
 
+    #[inline]
     pub fn duplicate(&self) -> io::Result<FileDesc> {
```
Comment (joshtriplett) [inline library/std/src/sys/unix/fd.rs:270]: Long-term, I wonder if we can eliminate this internal interface in favor of try_clone. But that's not an issue for this PR.

**Comment by joshtriplett** [comment]
LGTM. r=me when CI passes.

**Comment by sunfishcode** [comment]
The CI now passes; @joshtriplett I think this just needs an r+.

**Comment by joshtriplett** [comment]
LGTM

@bors r+

As a follow-up, I think we can drop the special-case handling of systems without F_DUPFD_CLOEXEC, as all our Linux targets require at least 2.6.32, and that flag was added in 2.6.24.

**Comment by Manishearth** [comment]
@bors r-

https://github.com/rust-lang/rust/pull/89525#issuecomment-933817867

**Comment by sunfishcode** [comment]
The compilation errors on WASI are now fixed.

**Comment by sunfishcode** [comment]
@joshtriplett This previously had a compilation error on WASI, which is now fixed.

**Comment by sunfishcode** [comment]
@joshtriplett This already had your r+, but it just needed an extra patch to fix the build. Would you mind taking a look?

**Comment by joshtriplett** [comment]
Sorry for the *long* delay re-approving this after the update.

@bors r+

**Comment by matthiaskrgr** [comment]
@bors r- 
failed in a rollup: https://github.com/rust-lang/rust/pull/91790#issuecomment-991692092

**Comment by sunfishcode** [comment]
The CI failure turned out to be that the code was calling `cvt`, and when it was moved to a different file, it silently started calling a different `cvt`. On Windows, one `cvt` treats 0 as indicating a failure, while the other `cvt` treats 0 as indicating success.

r? @joshtriplett

**Comment by joshtriplett** [comment]
@sunfishcode Good catch. That seems like a source of potential future bugs. Would you consider a separate PR to rename the socket-like one to `cvt_socket` or similar, to make the behavior difference more obvious?

@bors r+

**Pull request merged**